### ISSUE TITLE
Enforce integer sampling from multinomial

### ISSUE THREAD

**Issue by thalassemia** [CovertLab/wcEcoli#1399]
When [`numpy.random.RandomState.multinomial`](https://numpy.org/doc/stable/reference/random/generated/numpy.random.RandomState.multinomial.html) is called in an environment with `Cython<3.0.0` (currently, we use `0.29.34`), it accepts non-integer `n` values with no problem ([source code](https://github.com/numpy/numpy/blob/cdfbdf428d9df9c7119cecae323512a4cd3f57b7/numpy/random/mtrand.pyx#L4256) for reference). However, something changes in `Cython>=3.0.0` which causes the `multinomial` function to raise an error when `n` is not an integer. In our model, this happens in at least four places: [here](https://github.com/CovertLab/wcEcoli/blob/bd4eb473ac421406a83973986100ac3fe95fb55b/reconstruction/ecoli/initialization.py#L143C7-L143C7), [here](https://github.com/CovertLab/wcEcoli/blob/bd4eb473ac421406a83973986100ac3fe95fb55b/reconstruction/ecoli/initialization.py#L183), [here](https://github.com/CovertLab/wcEcoli/blob/bd4eb473ac421406a83973986100ac3fe95fb55b/models/ecoli/processes/rna_degradation.py#L363), and [here](https://github.com/CovertLab/wcEcoli/blob/bd4eb473ac421406a83973986100ac3fe95fb55b/models/ecoli/processes/rna_degradation.py#L422). The solution on our end is probably just to cast each of our `n` arguments to `int` once we upgrade Cython.

Digging deeper into the issue, I noticed that the `TypeError` for floating point values is only raised when a Cython function has a parameter of the type [`np.npy_intp`](https://numpy.org/doc/stable/reference/c-api/dtype.html#c.npy_intp). If the parameter has any other integer type (e.g. `int`, `long`, `np.npy_int`), it is simply cast to `int` with no errors raised. 

```cython
import numpy as np
cimport numpy as np
    
def func(int length):
    # func(1.1) returns 1
    return length

def weird_func(np.npy_intp length):
    # func(1.1) raises TypeError: 'float' object cannot be interpreted as an integer
    return length
```

Does anyone know if this issue is platform-specific (I tested on Ubuntu 22.04 x86) or intended behavior?

**Comment by 1fish2** [comment]
* Are you getting a `TypeError` when calling `multinomial()` from Python (as in those 4 places in wcEcoli) or only from Cython?

* Installing Cython 3.0.4 in Python 3.11.3 on macOS Ventura 13.6 (Intel CPU) then running the first example from the Numpy doc, except with a non-integer _n_:
  `    np.random.multinomial(20.49, [1/6.]*6, size=1)`
  _did not_ raise an exception. (Of course the environment's Cython should not affect Numpy -- pip should either be installing a pre-compiled Numpy or compiling it in a temporary virtual environment.)

* Compiling the `weird_func()` Cython example _did_ reproduce the exception on macOS:
  > TypeError: 'float' object cannot be interpreted as an integer

* I started looking through the monster Cython 3.0.0 unified release notes. Some key points may be buried there. This one matters and is a good step:  
  > The NumPy declarations (`cimport numpy`) were moved over to the NumPy project in order to allow version specific changes on their side.

* We'll have to ask whether this is a Cython bug or a Numpy bug.

* Note: Python 3.12 removes the `distutils` package, so we should sort out Cython 3.* issues and `setup.py` instructions before going to Python 3.12.

**Comment by thalassemia** [comment]
Good catch. I only see the issue when calling from Python. I have not tried calling from Cython. In my testing, the `TypeError` appears starting with [Numpy 1.26](generate_processes_and_steps). Since this is the first Numpy release with official support for Cython 3.0.0, I would not be surprised if the Numpy maintainers also used it to build the wheels starting with that release.

**Comment by 1fish2** [comment]
* I'll make wcEcoli work around this. See branch `cython_3`, which is a work in progress.
* The [doc for multinomial's](https://numpy.org/doc/stable/reference/random/generated/numpy.random.RandomState.multinomial.html) first argument says `n: int`. So rejecting a float isn't silly.
* I filed [a Numpy bug report](https://github.com/numpy/numpy/issues/25061) in case they want to change it back and in case they'll identify other APIs that changed and maybe we should adapt to.

I think the cause of this API change was a changed definition of `np.npy_intp`:
1. Numpy 1.26.0 took over managing their own, version-specific API bindings rather than having them embedded in Cython and subject to version skew.
2. That shifted the definition of `np.npy_intp` from `int` in Cython's [UFuncs.pyx](https://github.com/cython/cython/blob/bc32b7ec3f15f532c0ae59b930544c23172e8c50/Cython/Utility/UFuncs.pyx#L4) to `Py_ssize_t` in NumPy's [\_\_init\_\_.cython-30.pxd](https://github.com/numpy/numpy/blame/22ab9aa73cd80d13c74b8a227d7e3f10b84ab7fc/numpy/__init__.cython-30.pxd#L27).

**Comment by 1fish2** [comment]
The [numpy bug](https://github.com/numpy/numpy/issues/25061) is tagged for the [1.26.2 release milestone](https://github.com/numpy/numpy/milestone/123). If they fix it by the time we're ready for a new NumPy release, we won't need the `int()` workarounds although we'll still need the `np.product()` -> `np.prod()` update.

**Comment by thalassemia** [comment]
Thanks for linking the bug report you made! I find this kind of low-level stuff pretty fascinating. I cloned the Numpy repo and edited `__init__.cython-30.pxd` to define `npy_intp` as `int` instead of `Py_ssize_t` (this was on the `main` branch for Numpy 2.0). Building and installing allows me to do `np.random.multinomial(20.0, [1/6.]*6, size=1)` with no error. Changing the line back to `Py_ssize_t` and reinstalling causes the function call to return a `TypeError` again.

**Comment by thalassemia** [comment]
I think I've teased out all the details here. As noted by the [Cython maintainer](https://groups.google.com/g/cython-users/c/SLAKc1TEFww/m/iqZcb0b9BAAJ), `Py_ssize_t` uses an index-specific conversion protocol that is different from that used for other integer types. This protocol calls the `PyNumber_Index` function [here](https://github.com/cython/cython/blob/042451402cbbedbb3f6fc1b6324de880831367ff/Cython/Utility/TypeConversion.c#L509). This `PyNumber_Index` function calls `PyIndex_Check` [here](https://github.com/python/cpython/blob/11e83488c5a4a6e75a4f363a2e1a45574fd53573/Objects/abstract.c#L1478C14-L1478C14), and `PyIndex_Check` enforces the integer type [here](https://github.com/python/cpython/blob/f115a55f0e455a4b43a1da9fd838a60a101f182a/Include/internal/pycore_abstract.h#L16C6-L16C6) (`tp_as_number->nb_index` is a null pointer for `float` objects). 

Interestingly, commenting out the `PyNumber_Index` line in the generated Cython code and manually recompiling still leads to a `TypeError` being raised in [the subsequent call](https://github.com/cython/cython/blob/042451402cbbedbb3f6fc1b6324de880831367ff/Cython/Utility/TypeConversion.c#L511C17-L511C17) to `PyInt_AsSsize_t`. For Python >3.0, `PyInt_AsSsize_t` is defined to be `PyLong_AsSsize_t`, which enforces the integer type [here](https://github.com/python/cpython/blob/11e83488c5a4a6e75a4f363a2e1a45574fd53573/Objects/longobject.c#L579C19-L579C19).

Conversely, defining a parameter as `int` generates Cython code that calls `__Pyx_PyNumber_IntOrLong` [here](https://github.com/cython/cython/blob/042451402cbbedbb3f6fc1b6324de880831367ff/Cython/Utility/TypeConversion.c#L1190). This function converts `float` objects to `int` [here](https://github.com/cython/cython/blob/042451402cbbedbb3f6fc1b6324de880831367ff/Cython/Utility/TypeConversion.c#L449) (the `nb_int` slot of `float` objects is defined [here](https://github.com/python/cpython/blob/11e83488c5a4a6e75a4f363a2e1a45574fd53573/Objects/floatobject.c#L1868C14-L1868C14)).

This all aligns with what you said about Python only enforcing integer types for indices, exactly what the `Py_ssize_t` type was likely designed for. As for why Numpy is using an index type for `multinomial`, I found a [very old bug fix](https://github.com/charris/numpy/commit/862cc3b128e252294d7714cc89f638d2c5ed0c87) that changed many variables in the `mtrand` module to `npy_intp` since they were being used as indices. In the `multinomial` function, [this variable declaration](https://github.com/charris/numpy/blob/862cc3b128e252294d7714cc89f638d2c5ed0c87/numpy/random/mtrand/mtrand.pyx#L4071C10-L4071C10) was changed to `npy_intp`, forcing the input type to change as well to perform [this assignment](https://github.com/charris/numpy/blob/862cc3b128e252294d7714cc89f638d2c5ed0c87/numpy/random/mtrand/mtrand.pyx#L4094). To fix our issue, we can declare `dn` on a separate line as `long` since it is not used as a list index anywhere.

**Comment by 1fish2** [comment]
[NumPy 1.26.2](https://github.com/numpy/numpy/releases) has @thalassemia's bug fix and [the broader fix](https://github.com/numpy/numpy/pull/25095) for `Py_intptr_t` integer conversion.

I'll test it soon.
